core/src/main/java/org/apache/druid/crypto/CryptoService.java

```diff
@@ -0,0 +1,215 @@
+import java.nio.ByteBuffer;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.InvalidParameterSpecException;
+import java.security.spec.KeySpec;
+
+/**
+ * Utility class for symmetric key encryption (i.e. same secret is used for encryption and decryption) of byte[]
+ * using javax.crypto package.
+ *
+ * To learn about possible algorithms supported and their names,
+ * See https://docs.oracle.com/javase/8/docs/technotes/guides/security/StandardNames.html
+ */
+public class CryptoService
+{
+  // User provided secret phrase used for encrypting data
+  private final char[] passPhrase;
+
+  // Variables for algorithm used to generate a SecretKey based on user provided passPhrase
+  private final String secretKeyFactoryAlg;
+  private final int saltSize;
+  private final int iterationCount;
+  private final int keyLength;
+
+  // Cipher algorithm information
+  private final String cipherAlgName;
+  private final String cipherAlgMode;
+  private final String cipherAlgPadding;
+
+  // transformation =  "cipherAlgName/cipherAlgMode/cipherAlgPadding" used in Cipher.getInstance(transformation)
+  private final String transformation;
+
+  public CryptoService(
+      String passPhrase,
+      @Nullable String cipherAlgName,
+      @Nullable String cipherAlgMode,
+      @Nullable String cipherAlgPadding,
+      @Nullable String secretKeyFactoryAlg,
+      @Nullable Integer saltSize,
+      @Nullable Integer iterationCount,
+      @Nullable Integer keyLength
+  )
+  {
+    Preconditions.checkArgument(
+        passPhrase != null && !passPhrase.isEmpty(),
+        "null/empty passPhrase"
+    );
+    this.passPhrase = passPhrase.toCharArray();
+
+    this.cipherAlgName = cipherAlgName == null ? "AES" : cipherAlgName;
+    this.cipherAlgMode = cipherAlgMode == null ? "CBC" : cipherAlgMode;
+    this.cipherAlgPadding = cipherAlgPadding == null ? "PKCS5Padding" : cipherAlgPadding;
+    this.transformation = StringUtils.format("%s/%s/%s", this.cipherAlgName, this.cipherAlgMode, this.cipherAlgPadding);
+
+    this.secretKeyFactoryAlg = secretKeyFactoryAlg == null ? "PBKDF2WithHmacSHA256" : secretKeyFactoryAlg;
+    this.saltSize = saltSize == null ? 8 : saltSize;
+    this.iterationCount = iterationCount == null ? 65536 : iterationCount;
+    this.keyLength = keyLength == null ? 128 : keyLength;
+
+    // encrypt/decrypt a test string to ensure all params are valid
+    String testString = "duh! !! !!!";
+    Preconditions.checkState(
+        testString.equals(StringUtils.fromUtf8(decrypt(encrypt(StringUtils.toUtf8(testString))))),
+        "decrypt(encrypt(testString)) failed"
+    );
+  }
+
+  public byte[] encrypt(byte[] plain)
+  {
+    try {
+      byte[] salt = new byte[saltSize];
+      SecureRandom rnd = new SecureRandom();
```

Comment:
Could the SecureRandom here be static instead?

Reply:
yeah, changed